### ISSUE TITLE
Import csv into details

### ISSUE THREAD

**Issue by noeldum** [jam-py/jam-py#217]
I am adapting the code form the [export_import_csv_file howto](https://jam-py.com/docs/how_to/export_import_csv_files.html) to populate a csv file in the detail table of a selected record. As opposed to populate in the main table directly.

I made a dummy project having recipes in the journal and ingredients as details. My csv file contains the ingredients that I want to import in a selected recipe.

Something like that

![Screenshot_2021-08-28_12-48-03](https://user-images.githubusercontent.com/1875345/131216846-75af4069-4d1a-43e4-83c3-776e9a31e92f.png)

This is my code in client and server side for the journal table 'recipe'.

```
function on_view_form_created(item) {
  var csv_import_btn = item.add_view_button('Import CSV File');
  csv_import_btn.click(function() { csv_import(item) });
}

function csv_import(item) {
    var selected = item.task.recipe.selections;
    alert(item.task.recipe.selections);
    task.upload('static/files', {accept: '.csv', callback: function(file_name) {
        item.server('import_scv', [file_name, selected], function(error) {
            if (error) {
                item.warning(error);
            }
            item.refresh_page(true);
        });
    }});
}
```

```

 import os
 import csv
 
 def import_scv(item, file_name, selected):
    
    item.set_where(id__in=selected)
    print(selected)
    #copy = item.copy()
    copy = item.task.ingredients.copy()
    
    path = os.path.join(item.task.work_dir, 'static', 'files', file_name)
    with open(path, 'r', encoding='utf-8') as csvfile:
        copy.open(open_empty=True)
        reader = csv.DictReader(csvfile)
        for row in reader:
            print(row)
            copy.append()
            for field in copy.fields:
                if not field.system_field():
                    field.text = row[field.field_name]
            copy.post()
        copy.apply()
```

I am aware I need to inform the selection and pass it to the server. But somehow in js my task.recipe.selections does not return anything. I also played with rec_no but no luck.

Basically in the current state my code does not output any errors but clearly the csv does not import.

What is the correct way to inform the selection to the server side? 
Any general pointer for to accomplish what I am trying to do here?

jam-py 113 Python 3.8.10

**Comment by platipusica** [comment]
Is this working:
var selected = item.task.recipe.selections;

?
Because when you make recipe as an multi selection, than below works (in demo Invoices):

```
function csv_import(item) {
    //var selected = item.task.recipe.selections;
    
    var copy = item.copy({handlers: false}),
        selections = item.selections;
    if (selections.length > 1) {
        item.alert('Too many records selected.');
    }    
```
.
.
print(selected)
[464]

Of course, we can select more than one Invoice and your file upload will show...

**Comment by platipusica** [comment]
Btw, you need to study the definition of details table in Jam for full functionality.
The MASTER_ID and MASTER_REC_ID is missing due to:
 if not field.system_field():

Which is a bugger. @jam-py can you advice on this please?

thx

**Comment by noeldum** [comment]
The equivalent on the demo is if I wanted to import the tracks inside an invoice for example. All the purchased tracks being inside the csv file.

Well maybe selections is not what I need. Basically how do I tell the server side the selected entry for which i want to populate the details table?

**Comment by noeldum** [comment]
Right that selected entry is probably solved now using item.rec_no instead. My new code below.

```
function on_view_form_created(item) {
  var csv_import_btn = item.add_view_button('Import CSV File');
  csv_import_btn.click(function() { csv_import(item) });
}

function csv_import(item) {
    task.upload('static/files', {accept: '.csv', callback: function(file_name) {
        item.server('import_scv', [file_name, item.rec_no], function(error) {
            if (error) {
                item.warning(error);
            }
            item.refresh_page(true);
        });
    }});
}
```
```
import os
import csv

def import_scv(item, file_name, rec_no):
    
    item.set_where(id__in=rec_no)
    #copy = item.copy()
    copy = item.task.ingredients.copy()
    
    path = os.path.join(item.task.work_dir, 'static', 'files', file_name)
    with open(path, 'r', encoding='utf-8') as csvfile:
        copy.open(open_empty=True)
        reader = csv.DictReader(csvfile)
        for row in reader:
            print(row)
            copy.append()
            for field in copy.fields:
                if not field.system_field():
                    field.text = row[field.field_name]
            copy.post()
        copy.apply()
```

Now my problem is with the MASTER_ID and MASTER_REC_ID which are not informed. Thank you for pointing that out @platipusica. My entry are in the database but the 2 system field are NULL.

I thought the append method would take care of that.

**Comment by noeldum** [comment]
Got it to work somehow. Although I feel I am manipulating parameters that are probably abstracted. Let me know if there is a more elegant way than what I am doing. Here is my working code.

```
function on_view_form_created(item) {
  var csv_import_btn = item.add_view_button('Import CSV File');
  csv_import_btn.click(function() { csv_import(item) });
}

function csv_import(item) {
   // alert();
    task.upload('static/files', {accept: '.csv', callback: function(file_name) {
        item.server('import_scv', [file_name, item.rec_no], function(error) {
            if (error) {
                item.warning(error);
            }
            item.refresh_page(true);
        });
    }});
}
```
```

import os
import csv

def import_scv(item, file_name, rec_no):
    
    item.set_where(id__in=rec_no)
    #copy = item.copy()
    copy = item.task.ingredients.copy()
    
    path = os.path.join(item.task.work_dir, 'static', 'files', file_name)
    with open(path, 'r', encoding='utf-8') as csvfile:
        copy.open(open_empty=True)
        reader = csv.DictReader(csvfile)
        for row in reader:
            print(row)
            copy.append()
            for field in copy.fields:
                if not field.system_field():
                    field.text = row[field.field_name]
                else:
                    if field.field_name == 'master_rec_id':
                        field.text = rec_no+1
                    elif field.field_name == 'master_id':
                        field.text = item.ID
                        
            copy.post()
        copy.apply()
```

**Comment by platipusica** [comment]
👍

**Comment by platipusica** [comment]
Still, 

better @jam-py to advice since the Invoices/Details has Customer field, so when Importing CSV, the customer_id must be set, however this info is actually linked with Invoice Master from where we calling Import anyway. So in practise no need to set Customer_id in CSV, but...

Also, Import will fail due to Album and Artist id which must exist in CSV, ie:
[import.csv](https://github.com/jam-py/jam-py/files/7072474/import.csv)

This can be observed on vid. Again, Invoices must be set as Multi select. The unit price is 100.99 so we see that easily and where the problem is.

https://user-images.githubusercontent.com/9026100/131250361-422c4a08-6495-46ed-8b38-b0f480f10e23.mp4

**Comment by noeldum** [comment]
Actually my last code has a bug. How do I get the record ID of the selected entry on the table?

I figured the item.rec_no in my javascript is not the actual record id but just the number in the list. My code works at the moment because they match but it might not be always the case.

From both client and server side how do I get access to the record id of the selected entry?

**Comment by noeldum** [comment]
Answering my own question. The item must be opened before having access to the id. (arf getting there).
item.open()
print(item.id.value)

**Comment by noeldum** [comment]
I have solved that bug with the record id now. Here is my new code below.

```
function on_view_form_created(item) {
  var csv_import_btn = item.add_view_button('Import CSV File');
  csv_import_btn.click(function() { csv_import(item) });
}

function csv_import(item) {
    var rec_id = item.id.value;
    task.upload('static/files', {accept: '.csv', callback: function(file_name) {
        item.server('import_scv', [file_name, rec_id], function(error) {
            if (error) {
                item.warning(error);
            }
            item.refresh_page(true);
        });
    }});
}
```

```
import os
import csv

def import_scv(item, file_name, rec_id):
    
    copy = item.task.ingredients.copy()
    
    path = os.path.join(item.task.work_dir, 'static', 'files', file_name)
    with open(path, 'r', encoding='utf-8') as csvfile:
        copy.open(open_empty=True)
        reader = csv.DictReader(csvfile)
        for row in reader:
            print(row)
            copy.append()
            for field in copy.fields:
                if not field.system_field():
                    field.text = row[field.field_name]
                else:
                    if field.field_name == 'master_rec_id':
                        field.text = rec_id
                    elif field.field_name == 'master_id':
                        field.text = item.ID
                        
            copy.post()
        copy.apply()
```

Something I am dubious with my code above.
* I have not managed to get the record id directly from the item on the server side. Hence why I take it in rec_id on the client side and pass it through as argument in the import_scv function. item.id.value is always returning 0 here I don't know why.
* Not sure if I am right informing explicitly master_rec_id and master_id. Should that not be abstracted and managed differently?

Nonetheless that code above works ok the way it is although I am not convinced it's done in the state of art.

**Comment by platipusica** [comment]
@jam-py ?

**Comment by noeldum** [comment]
closing this as not really a bug or improvement topic. Will import that into the mailgroup at some point. 

Any comment welcome obviously.
